### PR TITLE
Creation of components through MLContext, internalization, and renaming

### LINKED ISSUE #1798: Creation of components through MLContext: advanced options and other feedback
In our estimators and other similar components often have advanced settings, because, sometimes people have unusual circumstances. At the same time, there is a 95% or 99% scenario for "simple" usage that most people will be happy with. For this reason we have often made a distinction between common and advanced settings, as we see here.

https://github.com/dotnet/machinelearning/blob/cb37c7e7f1e1b29b5608a2755db793c5435d10b1/src/Microsoft.ML.FastTree/TreeTrainersCatalog.cs#L29-L38

There are some possible things that excited feedback:

1. Echoing feedback seen in #1690, these things where we're making something should have the prefix `Create`, even in situations where this a catalog where we are *always* creating. Note: `Create` preferred to `Make`.

2. The worth of ASP.NET style configuration was questioned (seen above as `Action<FastTreeRegressionTrainer.Arguments> advancedSettings`), e.g., there may not be much purpose in having a delegate. The older style where it just takes the `Arguments` *period* was preferred.

3. Having this `Arguments` object as a nested class the component being created was viewed as positive, but this would be more idiomatically called `Options` -- `Arguments` was a holdover name from when these were exclusively for command line arguments, but for the API this is not a great name. So while keeping the general structure of how they are placed currently, they should probably be renamed to `Options`.

4. It is good to have the convenience for the simple arguments, however, if we have both simple and advanced settings, we should not mix them but have instead two distinct constructors/extension methods. (E.g., in the above, we would have two methods, one that took the advanced options.) To do otherwise is to invite confusion about which "wins" if we have the setting set in both.

    * Note that phase setting "set in both," which suggests that these settings object should retain the "simpler" settings in them. This reinforces feedback elsewhere as seen [here](https://github.com/dotnet/machinelearning/pull/1352#discussion_r228319239).

5. If the simple arguments are totally sufficient, then there is no need to expose this `Arguments` class in hte public API. (For practical reasons relating to command line and entry-point usage, we still need to always have these `Arguments` objects, but if they serve no purpose for the API the class can simply be made internal.)

/cc @KrzysztofCwalina, @terrajobst , on whose feedback this list is primarily based, and who can correct me and provide clarification in case I misspoke.

### PR DESCRIPTION
**Pull request by artidoro**
Fixes #1798, #1758 

1. In this PR I internalize the constructors of `DnnImageFeaturizerEstimator`, `PriorTrainer`, and `RandomTrainer`. This required to add catalog extensions for these three trainers. Internalizing these constructor closes issue #1798. 
2. I also rename a few instances of `Arguments` to `Options` which closes issue #1758.
3. I internalized some fields across `ITransformers` and `IEstimators`.
4. I also moved the `Options` class in TensorflowTransform to the estimator.

### CODE REVIEW

File: src/Microsoft.ML.Transforms/CustomMappingTransformer.cs [hunk 1]
```diff
@@ -197,7 +197,7 @@ public sealed class CustomMappingEstimator<TSrc, TDst> : TrivialEstimator<Custom
         /// <param name="contractName">The contract name, used by ML.NET for loading the model. If <c>null</c> is specified, such a trained model would not be save-able.</param>
         /// <param name="inputSchemaDefinition">Additional parameters for schema mapping between <typeparamref name="TSrc"/> and input data.</param>
         /// <param name="outputSchemaDefinition">Additional parameters for schema mapping between <typeparamref name="TDst"/> and output data.</param>
-        public CustomMappingEstimator(IHostEnvironment env, Action<TSrc, TDst> mapAction, string contractName,
+        internal CustomMappingEstimator(IHostEnvironment env, Action<TSrc, TDst> mapAction, string contractName,
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Transforms/CustomMappingTransformer.cs]: >CustomMappingEstimator [](start = 17, length = 22)

We don't have any samples/ tests with it at all?

**Comment by Ivanidzo4ka** [comment]
>         public override SchemaShape GetOutputSchema(SchemaShape inputSchema)

you used to add summary comments in previous PRs

---
Refers to: src/Microsoft.ML.Transforms/CustomMappingTransformer.cs:207 in 35e93ef. [](commit_id = 35e93ef2e06f7ad3c406cd674fa53486615e5d55, deletion_comment = False)

**Comment by Ivanidzo4ka** [comment]
>         public RandomModelParameters(IHostEnvironment env, int seed)

I think we trying to close our current predictors till 1.0

---
Refers to: src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs:142 in 35e93ef. [](commit_id = 35e93ef2e06f7ad3c406cd674fa53486615e5d55, deletion_comment = False)

**Comment by Ivanidzo4ka** [comment]
>         public PriorModelParameters(IHostEnvironment env, float prob)

internal?

---
Refers to: src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs:379 in 35e93ef. [](commit_id = 35e93ef2e06f7ad3c406cd674fa53486615e5d55, deletion_comment = False)

**Comment by abgoswam** [comment]
> // Licensed to the .NET Foundation under one or more agreements.

no changes in this file.

---
Refers to: src/Microsoft.ML.Data/Prediction/CalibratorCatalog.cs:1 in 35e93ef. [](commit_id = 35e93ef2e06f7ad3c406cd674fa53486615e5d55, deletion_comment = True)

File: src/Microsoft.ML.OnnxTransform/OnnxCatalog.cs [hunk 2]
```diff
@@ -61,5 +62,19 @@ public static class OnnxCatalog
...
+        /// <summary>
+        /// Applies a pre-trained DNN model to featurize an image.
+        /// </summary>
+        /// <param name="catalog">The transform's catalog.</param>
+        /// <param name="outputColumnName">The name of the column resulting from the transformation of <paramref name="inputColumnName"/>.</param>
+        /// <param name="modelFactory">An extension method on the <see cref="DnnImageModelSelector"/> that creates a chain of two
+        /// <see cref="OnnxScoringEstimator"/> (one for preprocessing and one with a pretrained image DNN) with specific models
+        /// included in a package together with that extension method.</param>
+        /// <param name="inputColumnName">Name of column to transform. If set to <see langword="null"/>, the value of the <paramref name="outputColumnName"/> will be used as source.</param>
+        public static DnnImageFeaturizerEstimator DnnFeaturizeImage(this TransformsCatalog catalog,
```
Comment (abgoswam) [inline src/Microsoft.ML.OnnxTransform/OnnxCatalog.cs]: >DnnFeaturizeImage [](start = 50, length = 17)

any examples for this in our samples ?

File: src/Microsoft.ML.StandardLearners/StandardLearnersCatalog.cs [hunk 3]
```diff
@@ -556,5 +556,29 @@ public static PoissonRegression PoissonRegression(this RegressionCatalog.Regress
 
             return new LinearSvmTrainer(CatalogUtils.GetEnvironment(catalog), options);
         }
+
+        /// <summary>
+        /// Predict a target using a random binary classification model <see cref="RandomTrainer"/>.
```
Comment (abgoswam) [inline src/Microsoft.ML.StandardLearners/StandardLearnersCatalog.cs]: ><see cref="RandomTrainer"/> [](start = 72, length = 27)

is sentence complete ?

File: src/Microsoft.ML.StandardLearners/StandardLearnersCatalog.cs [hunk 4]
```diff
@@ -556,5 +556,29 @@ public static PoissonRegression PoissonRegression(this RegressionCatalog.Regress
...
+            return new RandomTrainer(CatalogUtils.GetEnvironment(catalog), new RandomTrainer.Options());
+        }
+
+        /// <summary>
+        /// Predict a target using a binary classification model trained with <see cref="PriorTrainer"/> trainer.
+        /// </summary>
+        /// <param name="catalog">The <see cref="BinaryClassificationCatalog"/>.</param>
+        /// <param name="labelColumn">The name of the label column. </param>
+        /// <param name="weightsColumn">The optional name of the weights column.</param>
+        public static PriorTrainer Prior(this BinaryClassificationCatalog.BinaryClassificationTrainers catalog,
```
Comment (abgoswam) [inline src/Microsoft.ML.StandardLearners/StandardLearnersCatalog.cs]: >Prior [](start = 35, length = 5)

any samples we can link to ?

File: src/Microsoft.ML.Transforms/CustomMappingTransformer.cs (hunk 1 above)
Comment (artidoro) [reply src/Microsoft.ML.Transforms/CustomMappingTransformer.cs]: I am adding a sample now. 

---
In reply to: [255816087](https://github.com/dotnet/machinelearning/pull/2510#discussion_r255816087) [](ancestors = 255816087)

**Comment by artidoro** [comment]
>         public override SchemaShape GetOutputSchema(SchemaShape inputSchema)

Yes you are right let me do that.

---
In reply to: [462631911](https://github.com/dotnet/machinelearning/pull/2510#issuecomment-462631911) [](ancestors = 462631911)

---
Refers to: src/Microsoft.ML.Transforms/CustomMappingTransformer.cs:207 in 35e93ef. [](commit_id = 35e93ef2e06f7ad3c406cd674fa53486615e5d55, deletion_comment = False)

File: src/Microsoft.ML.OnnxTransform/OnnxCatalog.cs (hunk 2 above)
Comment (artidoro) [reply src/Microsoft.ML.OnnxTransform/OnnxCatalog.cs]: We don't have any unfortunately. We should add it, but don't think of doing this as part of this PR. Will make sure it's in the documentation issue though.

---
In reply to: [256071655](https://github.com/dotnet/machinelearning/pull/2510#discussion_r256071655) [](ancestors = 256071655)

File: src/Microsoft.ML.StandardLearners/StandardLearnersCatalog.cs (hunk 4 above)
Comment (artidoro) [reply src/Microsoft.ML.StandardLearners/StandardLearnersCatalog.cs]: I'll make a quick one for this and the random trainer. 

---
In reply to: [256072896](https://github.com/dotnet/machinelearning/pull/2510#discussion_r256072896) [](ancestors = 256072896)

File: src/Microsoft.ML.StandardLearners/StandardLearnersCatalog.cs (hunk 3 above)
Comment (artidoro) [reply src/Microsoft.ML.StandardLearners/StandardLearnersCatalog.cs]: I'll try to improve it.

---
In reply to: [256072727](https://github.com/dotnet/machinelearning/pull/2510#discussion_r256072727) [](ancestors = 256072727)

**Comment by artidoro** [comment]
>         public RandomModelParameters(IHostEnvironment env, int seed)

Ok will make this internal thank you!

---
In reply to: [462635433](https://github.com/dotnet/machinelearning/pull/2510#issuecomment-462635433) [](ancestors = 462635433)

---
Refers to: src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs:142 in 35e93ef. [](commit_id = 35e93ef2e06f7ad3c406cd674fa53486615e5d55, deletion_comment = False)

**Comment by artidoro** [comment]
>         public PriorModelParameters(IHostEnvironment env, float prob)

Thank you!


---
In reply to: [462635562](https://github.com/dotnet/machinelearning/pull/2510#issuecomment-462635562) [](ancestors = 462635562)

---
Refers to: src/Microsoft.ML.StandardLearners/Standard/Simple/SimpleTrainers.cs:379 in 35e93ef. [](commit_id = 35e93ef2e06f7ad3c406cd674fa53486615e5d55, deletion_comment = False)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/RandomTrainerSample.cs [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+            var trainedPipeline = pipeline.Fit(train);
+
+            // Step 4: Evaluate on the test set
+            var transformedData = trainedPipeline.Transform(test);
+            var evalMetrics = mlContext.BinaryClassification.Evaluate(transformedData, label: "Sentiment");
+
+            // Step 5: Inspect the output
+            Console.WriteLine("Accuracy: " + evalMetrics.Accuracy);
+
+            // Expected output (close to 0.5):
```
Comment (abgoswam) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/RandomTrainerSample.cs]: >(close to 0.5) [](start = 31, length = 14)

why is expected output close to 0.5 ...   does "Trainers.Random" mean  the prediction is random   or  does it mean  the  trainer is  selected randomly.  if its the former  than 0.5 makes sense.  but it its the later  than  i would expect  even our  most simple learners for binary classification are quite powerful .. thoughts ?

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/PriorTrainerSample.cs [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+            var trainedPipeline = pipeline.Fit(train);
+
+            // Step 4: Evaluate on the test set
+            var transformedData = trainedPipeline.Transform(test);
+            var evalMetrics = mlContext.BinaryClassification.Evaluate(transformedData, label: "Sentiment");
+
+            // Step 5: Inspect the output
+            Console.WriteLine("Accuracy: " + evalMetrics.Accuracy);
+
+            // Expected output:
```
Comment (abgoswam) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/PriorTrainerSample.cs]: >// Expected output: [](start = 12, length = 19)

am curious.. is there a 65-35 split in the dataset ?

Comment (artidoro) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/PriorTrainerSample.cs]: Yes that's what that means! Essentially it takes the proportion of a label in the training data set and outputs that as the probability of that label.

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/RandomTrainerSample.cs (hunk 5 above)
Comment (artidoro) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/RandomTrainerSample.cs]: We expect something close to 0.5 because this trainer outputs a random prediction, or in other words a label at random regardless of the input features. In fact, it does not take any feature column. 

We are not taking a trainer in ML.NET at random. 

I will try to make this a bit clearer. Thanks for asking!

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/RandomTrainerSample.cs [hunk 7]
```diff
@@ -0,0 +1,67 @@
+﻿using System;
+using System.Linq;
+using Microsoft.ML.Data;
+using Microsoft.ML.Trainers;
+
+namespace Microsoft.ML.Samples.Dynamic
+{
+    public class RandomTrainerSample
```
Comment (Ivanidzo4ka) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/RandomTrainerSample.cs]: >RandomTrainerSample [](start = 17, length = 19)

Zeeshan A, Shahab and me have PRs where we create BinaryClassification folder.
Any chance you can move this two to that folder?
